File: app/src/main/java/com/tablet/hid/ui/community/GamepadThumbnailView.kt
package com.tablet.hid.ui.community

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.RectF
import android.util.AttributeSet
import android.view.View
import com.tablet.hid.util.GamepadElementRect
import com.tablet.hid.util.GamepadLayoutResolver
import org.json.JSONObject

class GamepadThumbnailView @JvmOverloads constructor(
    context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0,
) : View(context, attrs, defStyleAttr) {

    companion object {
        private const val REF_LONG = 600f
        private const val REF_SHORT = 340f
        private val BUTTON_COLORS = mapOf(
            "a" to 0xFF4ade80.toInt(),
            "b" to 0xFFf87171.toInt(),
            "x" to 0xFF60a5fa.toInt(),
            "y" to 0xFFfbbf24.toInt(),
        )
        private const val NEUTRAL_COLOR = 0xFFb0b8c1.toInt()
        private val BUTTON_KEYS = listOf(
            "a", "b", "x", "y", "lb", "rb", "lt", "rt",
            "back", "start", "dpadUp", "dpadDown", "dpadLeft", "dpadRight",
        )
    }

    private var isLandscape = true
    private var layout: Map<String, GamepadElementRect> = emptyMap()
    private val buttonEnabled = mutableMapOf<String, Boolean>()
    private val buttonOX = mutableMapOf<String, Float>()
    private val buttonOY = mutableMapOf<String, Float>()
    private val buttonSX = mutableMapOf<String, Float>()
    private val buttonSY = mutableMapOf<String, Float>()
    private var leftJoyEnabled = true
    private var rightJoyEnabled = true
    private var singleMode = false
    private var leftOX = 0f; private var leftOY = 0f; private var leftSX = 1f
    private var rightOX = 0f; private var rightOY = 0f; private var rightSX = 1f

    private val fillPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply { style = Paint.Style.FILL }
    private val strokePaint = Paint(Paint.ANTI_ALIAS_FLAG).apply { style = Paint.Style.STROKE; strokeWidth = 1.5f }
    private val bgPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply { style = Paint.Style.FILL; color = 0xFF0f0f1a.toInt() }
    private val rect = RectF()

    fun setLandscape(landscape: Boolean) {
        if (isLandscape == landscape) return
        isLandscape = landscape
        layout = emptyMap()
        requestLayout()
        invalidate()
    }

    fun setConfigJson(json: String) {
        try {
            val root = JSONObject(json)
            val buttons = root.optJSONObject("buttons") ?: JSONObject()
            for (key in BUTTON_KEYS) {
                val btn = buttons.optJSONObject(key)
                buttonEnabled[key] = btn?.optBoolean("enabled", true) ?: true
                buttonOX[key] = btn?.optDouble("offsetX", 0.0)?.toFloat() ?: 0f
                buttonOY[key] = btn?.optDouble("offsetY", 0.0)?.toFloat() ?: 0f
                buttonSX[key] = btn?.optDouble("scaleX", 1.0)?.toFloat() ?: 1f
                buttonSY[key] = btn?.optDouble("scaleY", 1.0)?.toFloat() ?: 1f
            }
            val leftJs = root.optJSONObject("leftJoystick")
            val rightJs = root.optJSONObject("rightJoystick")
            leftJoyEnabled = leftJs?.optBoolean("enabled", true) ?: true
            rightJoyEnabled = rightJs?.optBoolean("enabled", true) ?: true
            singleMode = root.optBoolean("singleJoystickMode", false)
            leftOX = leftJs?.optDouble("offsetX", 0.0)?.toFloat() ?: 0f
            leftOY = leftJs?.optDouble("offsetY", 0.0)?.toFloat() ?: 0f
            leftSX = leftJs?.optDouble("scaleX", 1.0)?.toFloat() ?: 1f
            rightOX = rightJs?.optDouble("offsetX", 0.0)?.toFloat() ?: 0f
            rightOY = rightJs?.optDouble("offsetY", 0.0)?.toFloat() ?: 0f
            rightSX = rightJs?.optDouble("scaleX", 1.0)?.toFloat() ?: 1f
        } catch (_: Exception) { }
        layout = emptyMap()
        ensureLayout()
        invalidate()
    }

    private fun refW() = if (isLandscape) REF_LONG else REF_SHORT
    private fun refH() = if (isLandscape) REF_SHORT else REF_LONG

    private fun ensureLayout() {
        if (layout.isEmpty()) {
            layout = GamepadLayoutResolver.resolveLayout(context, refW(), refH())
        }
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        ensureLayout()
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val w = MeasureSpec.getSize(widthMeasureSpec).coerceAtLeast(1)
        val h = (w * refH() / refW()).toInt().coerceAtLeast(1)
        setMeasuredDimension(w, h)
    }

    override fun onDraw(canvas: Canvas) {
        ensureLayout()
        if (layout.isEmpty()) return
        val scale = width.toFloat() / refW()
        val cornerBg = 12f * scale

        canvas.drawRoundRect(0f, 0f, width.toFloat(), height.toFloat(), cornerBg, cornerBg, bgPaint)

        for (key in BUTTON_KEYS) {
            val nat = layout[key] ?: continue
            if (buttonEnabled[key] == false) continue
            drawButton(canvas, key, nat, scale)
        }

        val leftNat = layout["leftJoystick"]
        val rightNat = layout["rightJoystick"]
        if (leftJoyEnabled && leftNat != null) {
            drawJoystick(canvas, leftNat, leftOX, leftOY, leftSX, scale)
        }
        if (rightJoyEnabled && !singleMode && rightNat != null) {
            drawJoystick(canvas, rightNat, rightOX, rightOY, rightSX, scale)
        }
    }

    private fun drawButton(canvas: Canvas, key: String, nat: GamepadElementRect, scale: Float) {
        val ox = buttonOX[key] ?: 0f
        val oy = buttonOY[key] ?: 0f
        val sx = buttonSX[key] ?: 1f
        val sy = buttonSY[key] ?: 1f
        val cx = (nat.left + nat.w / 2f + ox) * scale
        val cy = (nat.top + nat.h / 2f + oy) * scale
        val hw = nat.w * sx * scale / 2f
        val hh = nat.h * sy * scale / 2f
        val baseColor = BUTTON_COLORS[key] ?: NEUTRAL_COLOR
        val isTrigger = key == "lt" || key == "rt"
        val isDpad = key.startsWith("dpad")
        val corner = when {
            isTrigger -> 4f * scale
            isDpad    -> 3f * scale
            else      -> 14f * scale
        }
        rect.set(cx - hw, cy - hh, cx + hw, cy + hh)
        fillPaint.color = (0x33 shl 24) or (baseColor and 0x00FFFFFF)
        canvas.drawRoundRect(rect, corner, corner, fillPaint)
        strokePaint.color = (0x99 shl 24) or (baseColor and 0x00FFFFFF)
        canvas.drawRoundRect(rect, corner, corner, strokePaint)
    }

    private fun drawJoystick(
        canvas: Canvas, nat: GamepadElementRect,
        ox: Float, oy: Float, sx: Float, scale: Float,
    ) {
        val cx = (nat.left + nat.w / 2f + ox) * scale
        val cy = (nat.top + nat.h / 2f + oy) * scale
        val radius = nat.w * sx * scale / 2f

        fillPaint.color = 0x1AFFFFFF
        canvas.drawCircle(cx, cy, radius, fillPaint)
        strokePaint.color = 0x60FFFFFF
        canvas.drawCircle(cx, cy, radius, strokePaint)

        val knobRadius = radius * 0.38f
        fillPaint.color = 0xBFFFFFFF.toInt()
        canvas.drawCircle(cx, cy, knobRadius, fillPaint)
    }
}

File: ios/TabletHID/Views/Community/GamepadThumbnailView.swift
import SwiftUI

/// Read-only preview of a gamepad layout, drawn onto a SwiftUI Canvas.
///
/// Mirrors Android's GamepadThumbnailView: reads gamepad_layout.json via GamepadLayoutResolver,
/// scales natural element rects to the view's pixel width, draws filled+stroked buttons and
/// joystick circles. The aspect ratio is always refH/refW (landscape or portrait).
struct GamepadThumbnailView: View {

    // MARK: - Configuration

    /// Resolved config to display. Mutually exclusive with configJson.
    var config: GamepadConfig?

    /// Raw canonical JSON string to display (used when importing without a local model).
    var configJson: String?

    /// True → 600×340 reference canvas (landscape). False → 340×600 (portrait).
    var isLandscape: Bool = true

    /// Optional source canvas long/short side (dp). When nil uses 600/340 defaults.
    var customCanvasLong: Double? = nil
    var customCanvasShort: Double? = nil

    // MARK: - Constants

    private static let refLong: Double = 600
    private static let refShort: Double = 340

    private static let buttonColors: [String: Color] = [
        "a": Color(red: 0x4a/255.0, green: 0xde/255.0, blue: 0x80/255.0),
        "b": Color(red: 0xf8/255.0, green: 0x71/255.0, blue: 0x71/255.0),
        "x": Color(red: 0x60/255.0, green: 0xa5/255.0, blue: 0xfa/255.0),
        "y": Color(red: 0xfb/255.0, green: 0xbf/255.0, blue: 0x24/255.0),
    ]
    private static let neutralColor = Color(red: 0xb0/255.0, green: 0xb8/255.0, blue: 0xc1/255.0)
    private static let bgColor      = Color(red: 0x0f/255.0, green: 0x0f/255.0, blue: 0x1a/255.0)

    private static let buttonKeys = [
        "a", "b", "x", "y", "lb", "rb", "lt", "rt",
        "back", "start", "dpadUp", "dpadDown", "dpadLeft", "dpadRight",
    ]

    // MARK: - Computed layout data

    private var refW: Double { isLandscape ? Self.refLong  : Self.refShort }
    private var refH: Double { isLandscape ? Self.refShort : Self.refLong  }

    private func effectiveRefW() -> Double {
        guard let l = customCanvasLong, let s = customCanvasShort else { return refW }
        return isLandscape ? l : s
    }
    private func effectiveRefH() -> Double {
        guard let l = customCanvasLong, let s = customCanvasShort else { return refH }
        return isLandscape ? s : l
    }

    // MARK: - Parsed state (computed from config/configJson)

    private struct LayoutState {
        var buttonEnabled  = [String: Bool]()
        var buttonOX       = [String: Double]()
        var buttonOY       = [String: Double]()
        var buttonSX       = [String: Double]()
        var buttonSY       = [String: Double]()
        var leftEnabled    = true
        var rightEnabled   = true
        var singleMode     = false
        var leftOX: Double = 0; var leftOY: Double = 0;  var leftSX:  Double = 1
        var rightOX: Double = 0; var rightOY: Double = 0; var rightSX: Double = 1
    }

    private var layoutState: LayoutState {
        var state = LayoutState()
        if let cfg = config {
            let buttons: [(String, ButtonConfig)] = [
                ("a", cfg.btnA), ("b", cfg.btnB), ("x", cfg.btnX), ("y", cfg.btnY),
                ("lb", cfg.btnLB), ("rb", cfg.btnRB), ("lt", cfg.btnLT), ("rt", cfg.btnRT),
                ("back", cfg.btnBack), ("start", cfg.btnStart),
                ("dpadUp", cfg.dpadUp), ("dpadDown", cfg.dpadDown),
                ("dpadLeft", cfg.dpadLeft), ("dpadRight", cfg.dpadRight),
            ]
            for (key, btn) in buttons {
                state.buttonEnabled[key] = btn.enabled
                state.buttonOX[key] = btn.offsetX
                state.buttonOY[key] = btn.offsetY
                state.buttonSX[key] = btn.scaleX
                state.buttonSY[key] = btn.scaleY
            }
            state.leftEnabled  = cfg.leftJoystick.enabled
            state.rightEnabled = cfg.rightJoystick.enabled
            state.singleMode   = cfg.singleJoystickMode
            state.leftOX  = cfg.leftJoystick.offsetX;  state.leftOY  = cfg.leftJoystick.offsetY
            state.leftSX  = cfg.leftJoystick.scaleX
            state.rightOX = cfg.rightJoystick.offsetX; state.rightOY = cfg.rightJoystick.offsetY
            state.rightSX = cfg.rightJoystick.scaleX
        } else if let jsonStr = configJson,
                  let data = jsonStr.data(using: .utf8),
                  let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        {
            let buttons = root["buttons"] as? [String: Any] ?? [:]
            for key in Self.buttonKeys {
                let btn = buttons[key] as? [String: Any]
                state.buttonEnabled[key] = btn?["enabled"] as? Bool ?? true
                state.buttonOX[key] = btn?["offsetX"] as? Double ?? 0
                state.buttonOY[key] = btn?["offsetY"] as? Double ?? 0
                state.buttonSX[key] = btn?["scaleX"]  as? Double ?? 1
                state.buttonSY[key] = btn?["scaleY"]  as? Double ?? 1
            }
            let leftJs  = root["leftJoystick"]  as? [String: Any]
            let rightJs = root["rightJoystick"] as? [String: Any]
            state.leftEnabled  = leftJs?["enabled"]  as? Bool ?? true
            state.rightEnabled = rightJs?["enabled"] as? Bool ?? true
            state.singleMode   = root["singleJoystickMode"] as? Bool ?? false
            state.leftOX  = leftJs?["offsetX"]  as? Double ?? 0
            state.leftOY  = leftJs?["offsetY"]  as? Double ?? 0
            state.leftSX  = leftJs?["scaleX"]   as? Double ?? 1
            state.rightOX = rightJs?["offsetX"] as? Double ?? 0
            state.rightOY = rightJs?["offsetY"] as? Double ?? 0
            state.rightSX = rightJs?["scaleX"]  as? Double ?? 1
        }
        return state
    }

    // MARK: - View body

    var body: some View {
        let rw = effectiveRefW()
        let rh = effectiveRefH()
        let layout = GamepadLayoutResolver.resolveLayout(canvasW: rw, canvasH: rh)
        let state  = layoutState

        GeometryReader { geo in
            let scale = geo.size.width / rw
            Canvas { ctx, size in
                // Background
                ctx.fill(
                    Path(roundedRect: CGRect(origin: .zero, size: size), cornerRadius: 12 * scale),
                    with: .color(Self.bgColor)
                )

                // Buttons
                for key in Self.buttonKeys {
                    guard state.buttonEnabled[key] != false,
                          let nat = layout[key] else { continue }
                    let ox = state.buttonOX[key] ?? 0
                    let oy = state.buttonOY[key] ?? 0
                    let sx = state.buttonSX[key] ?? 1
                    let sy = state.buttonSY[key] ?? 1
                    let cx  = (nat.left + nat.w / 2.0 + ox) * scale
                    let cy  = (nat.top  + nat.h / 2.0 + oy) * scale
                    let hw  = nat.w * sx * scale / 2.0
                    let hh  = nat.h * sy * scale / 2.0
                    let base = Self.buttonColors[key] ?? Self.neutralColor
                    let isTrigger = key == "lt" || key == "rt"
                    let isDpad    = key.hasPrefix("dpad")
                    let corner: Double = isTrigger ? 4 * scale : (isDpad ? 3 * scale : 14 * scale)
                    let rect = CGRect(x: cx - hw, y: cy - hh, width: hw * 2, height: hh * 2)
                    ctx.fill(
                        Path(roundedRect: rect, cornerRadius: corner),
                        with: .color(base.opacity(0.2))
                    )
                    ctx.stroke(
                        Path(roundedRect: rect, cornerRadius: corner),
                        with: .color(base.opacity(0.6)),
                        lineWidth: 1.5
                    )
                }

                // Left joystick
                if state.leftEnabled, let nat = layout["leftJoystick"] {
                    drawJoystick(ctx: ctx, nat: nat, ox: state.leftOX, oy: state.leftOY,
                                 sx: state.leftSX, scale: scale)
                }

                // Right joystick (hidden in single-mode)
                if state.rightEnabled && !state.singleMode, let nat = layout["rightJoystick"] {
                    drawJoystick(ctx: ctx, nat: nat, ox: state.rightOX, oy: state.rightOY,
                                 sx: state.rightSX, scale: scale)
                }
            }
        }
        .aspectRatio(rw / rh, contentMode: .fit)
        .clipShape(RoundedRectangle(cornerRadius: 12))
    }

    // MARK: - Drawing helper

    private func drawJoystick(
        ctx: GraphicsContext,
        nat: GamepadElementRect,
        ox: Double, oy: Double, sx: Double,
        scale: Double
    ) {
        let cx     = (nat.left + nat.w / 2.0 + ox) * scale
        let cy     = (nat.top  + nat.h / 2.0 + oy) * scale
        let radius = nat.w * sx * scale / 2.0
        let outer  = Path(ellipseIn: CGRect(x: cx - radius, y: cy - radius,
                                            width: radius * 2, height: radius * 2))
        ctx.fill(outer,   with: .color(.white.opacity(0.10)))
        ctx.stroke(outer, with: .color(.white.opacity(0.38)), lineWidth: 1.5)

        let kr     = radius * 0.38
        let knob   = Path(ellipseIn: CGRect(x: cx - kr, y: cy - kr, width: kr * 2, height: kr * 2))
        ctx.fill(knob, with: .color(.white.opacity(0.75)))
    }
}
